### PR TITLE
Adding / Deleting Profile Picture

### LINKED ISSUE #118: Set Group Thumbnail
Hello friends

Is there any way to set group thumbnail via WAPI call?

Thanks in advance

### PR DESCRIPTION
**Pull request by purpshell**
# Profile Picture functionality

##  INFO

This pull request is adding the way to set and delete profile pictures. 

Currently, this only works with chats in cache, (ProfilePicThumb), 
I will see if I can change that though..

It automatically resizes any unfitting dimensions, picking the left of an image.. Recommended dimensions are 640*640.

**YOU CAN ONLY SET / DELETE PFPs TO GROUPS and YOURSELF**
(you must be able to do it on your own by default)
(you must have group admin on any chat that has restricting editing info only to admins)

EDIT (May 4, 2021): THIS HAS BEEN TESTED FOR 7 DAYS.. All seems working with no bans and no potential risk of being banned.

## VIDEO DEMO:
https://user-images.githubusercontent.com/52720753/116287058-9f098180-a798-11eb-9ed4-657af6952d93.mp4

## USAGE

This pull request adds two methods. 

`client.setProfilePicture(id, picture)`: 
  **-** Parameters: 
    - id : String (Chat ID to set)
    - picture: MessageMedia instance of the image you want to be set.
  **-** Returns: 
  Image URL / `False`
`client.deleteProfilePicture(id)`:
  **-** Parameters: 
     - id : String (Chat ID to delete from)
  **-** Returns: 
  `True` / `False`
Also 2 sub-methods for a GroupChat instance: 

`chat.setProfilePicture(picture)`: 
  **-** Parameters: 
    - picture: MessageMedia instance of the image you want to be set.
  **-** Returns: 
  Image URL / `False`

`chat.deleteProfilePicture()`:
  **-** Parameters: 
     _none_
  **-** Returns: 
  `True` / `False`

### CODE REVIEW

**Review by wictornogueira — APPROVED** [review]

**Comment by guicuton** [comment]
When these methods will be merged?  They would be very usefull :-)

**Comment by bernardoely** [comment]
Looking forward to getting this functionality!! Really appreciate if you can merge this. Thanks.

File: src/structures/GroupChat.js [hunk 1]
```diff
@@ -75,6 +75,23 @@ class GroupChat extends Chat {
...
+    async setProfilePicture(picture) {
+        return await this.client.setProfilePicture(this.id._serialized, picture);
+    }
+
+    /**
+     * Deletes the profile picture of a group chat.
+     * @returns {Promise<Boolean>} true if it succeeded / false if it didn't delete the profile picture.
+     */
+    async deleteProfilePicture() {
+        return await this.client.setProfilePicture(this.id._serialized);
```
Comment (pedroslopez) [inline src/structures/GroupChat.js:92]: Should this be calling `deleteProfilePicture` instead of `setProfilePicture`?

**Review by pedroslopez — CHANGES_REQUESTED** [review]
This is great! Thanks for adding this.

I think to avoid confusion we should leave `client.setProfilePicture` and `client.removeProfilePicture` to deal with the current user's profile and let `groupChat.setProfilePicture` / `groupChat.removeProfilePicture` deal with the group's picture. You can then move common functionality to Injected.js

File: src/structures/GroupChat.js (hunk 1 above)
Comment (PurpShell) [reply src/structures/GroupChat.js:92]: oops

I usually make PRs locally and test then re-type it out in github's editor.. 

I am going to use git functions from now on, overall less mistakes

Comment (PurpShell) [reply src/structures/GroupChat.js:92]: Thanks for putting an eye on this, sorry for the confusion

**Comment by Murilandrade** [comment]
when I try to do that a I get an erro on the window.mR.findModule('setProfilePic')[0]; is there a solution for this?

**Comment by PurpShell** [comment]
> when I try to do that I get an erro on the window.mR.findModule('setProfilePic')[0]; is there a solution for this?

I will check it out. i still have to re-factor the code anyways so I will have to find the module by collection and not the method name itself 

Best regards, 

[PurpShell](https://github.com/PurpShell)

**Comment by PurpShell** [comment]
As you might have seen, the PR has been put on draft mode and the eslint failed

**Comment by arthur99eel** [comment]
@PurpShell great feature! I have some doubts on the usage, I realized the function formatImageToProfilePic is missing in the Util class, what formats this function has to return on the img and preview properties?

**Comment by bragarpaulo** [comment]
Very nice this!!!! I hope to use it soon.

Congrats @PurpShell & @pedroslopez

**Review by Gimenz — APPROVED** [review]

**Comment by eowoosh** [comment]
Hello! It's possible you share function Util.formatImageToProfilePic() ?

`TypeError: Util.formatImageToProfilePic is not a function`

**Comment by dbsdenis** [comment]
> @PurpShell Hello! É possível compartilhar a função Util.formatImageToProfilePic ()?
> 
> `TypeError: Util.formatImageToProfilePic is not a function`

`Exemplo de uso:`

> 
    if (msg.body === '!addpicture') {
        // Add picture the group
        let chat = await msg.getChat();
        let picture = MessageMedia.fromFilePath("image.png");
        if (chat.isGroup) {
            await client.setProfilePicture(chat.id._serialized, picture);
        } else {
            msg.reply('This command can only be used in a group!');
        }
    }

    if (msg.body === '!removepicture') {
        // Remove picture the group
        let chat = await msg.getChat();
        if (chat.isGroup) {
            await client.deleteProfilePicture(chat.id._serialized);
        } else {
            msg.reply('This command can only be used in a group!');
        }
    }


`src/util/Util.js`

> 
    /**
     * Formats a image to profile picture
     * @param {MessageMedia} media
     * 
     * @returns {Promise<MessageMedia>} media in picture format
     */
    static async formatImageToProfilePic(media) {

        if (!media.mimetype.includes('image'))
            throw new Error('media is not a image');

        const buff = Buffer.from(media.data, 'base64');

        let preview = sharp(buff).resize(96, 96);
        let img = sharp(buff).resize(640, 640);
      
        preview = `data:${media.mimetype};base64,${(await preview.toBuffer()).toString('base64')}`;
        img = `data:${media.mimetype};base64,${(await img.toBuffer()).toString('base64')}`;

        return {             
            img,
            preview
        };
    }

**Comment by PurpShell** [comment]
> > @PurpShell Hello! É possível compartilhar a função Util.formatImageToProfilePic ()?
> > `TypeError: Util.formatImageToProfilePic is not a function`
> 
> `Exemplo de uso:`
> 
> > 
> 
> ```
> if (msg.body === '!addpicture') {
>     // Add picture the group
>     let chat = await msg.getChat();
>     let picture = MessageMedia.fromFilePath("image.png");
>     if (chat.isGroup) {
>         await client.setProfilePicture(chat.id._serialized, picture);
>     } else {
>         msg.reply('This command can only be used in a group!');
>     }
> }
> 
> if (msg.body === '!removepicture') {
>     // Remove picture the group
>     let chat = await msg.getChat();
>     if (chat.isGroup) {
>         await client.deleteProfilePicture(chat.id._serialized);
>     } else {
>         msg.reply('This command can only be used in a group!');
>     }
> }
> ```
> 
> `src/util/Util.js`
> 
> > 
> 
> ```
> /**
>  * Formats a image to profile picture
>  * @param {MessageMedia} media
>  * 
>  * @returns {Promise<MessageMedia>} media in picture format
>  */
> static async formatImageToProfilePic(media) {
> 
>     if (!media.mimetype.includes('image'))
>         throw new Error('media is not a image');
> 
>     const buff = Buffer.from(media.data, 'base64');
> 
>     let preview = sharp(buff).resize(96, 96);
>     let img = sharp(buff).resize(640, 640);
>   
>     preview = `data:${media.mimetype};base64,${(await preview.toBuffer()).toString('base64')}`;
>     img = `data:${media.mimetype};base64,${(await img.toBuffer()).toString('base64')}`;
> 
>     return {             
>         img,
>         preview
>     };
> }
> ```

this is different from my way, though i haven't been able to push since i have been getting issues with `sharp`. 

also the way you do this is very bad, resizing an image of any dimension / aspect ratio, and also not providing the format in JPEG.

**Comment by PurpShell** [comment]
I am working on this pull request again! Sorry for the long wait! I'll work more on this on my windows machine, since no sharp issues there! So far I've tried to imitate the original way the function worked, but it seems like something has either changed or i'm getting it wrong. Excuse me for not providing the util function! I reset my `node_modules` a bunch of times since April, which caused me to lose the function.

**Comment by PurpShell** [comment]
Since I am getting rid of sharp, which this PR uses, I will also find the functions whatsapp uses to resize the image, and finally settle this issue

**Comment by DangerD1024** [comment]
Hi, when you will merge that? =)

**Pull request closed without merging**